state/protocol/protocol_state/updater_test.go

```diff
@@ -0,0 +1,464 @@
+		})
+
+		err = updater.ProcessEpochCommit(commit)
+		require.NoError(s.T(), err)
+
+		// processing another epoch commit has to be an error since we have already processed one
+		err = updater.ProcessEpochCommit(commit)
+		require.Error(s.T(), err)
+
+		newState, _, _ := updater.Build()
+		require.Equal(s.T(), commit.ID(), newState.NextEpochProtocolState.CurrentEpochEventIDs.CommitID, "next epoch must be committed")
+	})
+}
+
+// TestUpdateIdentityUnknownIdentity tests if updating the identity of unknown node results in an error.
+func (s *UpdaterSuite) TestUpdateIdentityUnknownIdentity() {
+	identity := &flow.DynamicIdentityEntry{
+		NodeID:  unittest.IdentifierFixture(),
+		Dynamic: flow.DynamicIdentity{},
+	}
+	err := s.updater.UpdateIdentity(identity)
+	require.Error(s.T(), err, "should not be able to update data of unknown identity")
+
+	_, _, hasChanges := s.updater.Build()
+	require.False(s.T(), hasChanges, "should not have changes")
+}
+
+// TestUpdateIdentityHappyPath tests if identity updates are correctly processed and reflected in the resulting protocol state.
+func (s *UpdaterSuite) TestUpdateIdentityHappyPath() {
+	// update protocol state to have next epoch protocol state
+	unittest.WithNextEpochProtocolState()(s.parentProtocolState)
+	s.updater = newUpdater(s.candidate, s.parentProtocolState)
+
+	currentEpochParticipants := s.parentProtocolState.CurrentEpochSetup.Participants.Copy()
+	weightChanges, err := currentEpochParticipants.Sample(2)
+	require.NoError(s.T(), err)
+	ejectedChanges, err := currentEpochParticipants.Sample(2)
+	require.NoError(s.T(), err)
+	for i, identity := range weightChanges {
+		identity.DynamicIdentity.Weight = uint64(100 * i)
+	}
+	for _, identity := range ejectedChanges {
+		identity.Ejected = true
+	}
+
+	allUpdates := append(weightChanges, ejectedChanges...)
+	for _, update := range allUpdates {
+		err := s.updater.UpdateIdentity(&flow.DynamicIdentityEntry{
+			NodeID:  update.NodeID,
+			Dynamic: update.DynamicIdentity,
+		})
+		require.NoError(s.T(), err)
+	}
+	updatedState, _, hasChanges := s.updater.Build()
+	require.True(s.T(), hasChanges, "should have changes")
+
+	requireUpdatesApplied := func(identityLookup map[flow.Identifier]*flow.DynamicIdentityEntry) {
+		for _, identity := range allUpdates {
+			updatedIdentity := identityLookup[identity.NodeID]
+			require.Equal(s.T(), identity.NodeID, updatedIdentity.NodeID)
+			require.Equal(s.T(), identity.DynamicIdentity, updatedIdentity.Dynamic, "identity should be updated")
+		}
+	}
+
+	// check if changes are reflected in current and next epochs
+	requireUpdatesApplied(updatedState.Identities.Lookup())
+	requireUpdatesApplied(updatedState.NextEpochProtocolState.Identities.Lookup())
+}
+
+// TestProcessEpochSetupInvariants tests if processing epoch setup when invariants are violated doesn't update internal structures.
+func (s *UpdaterSuite) TestProcessEpochSetupInvariants() {
```

Comment:
I feel we should also validate that `TransitionToNextEpoch()` errors when `InvalidStateTransitionAttempted == true`. I think currently, the implementation of `TransitionToNextEpoch()` does not really cover this case explicitly  (?)

Reply:
https://github.com/onflow/flow-go/pull/4579/commits/d222fd0a645ebd03a8c27df26857cee47c354689